src/main/java/hudson/tasks/test/AbstractTestResultAction.java

```diff
@@ -287,29 +350,249 @@ public TestResult findCorrespondingResult(String id) {
         return Collections.emptyList();
     }
 
+    /**
+     * Start for Test Result Trends.
+     * From here onwards start the code area responsible for generating various test result trends.
+     */
+
+    public hudson.tasks.junit.TestResult loadXml() {
+        return null;
+    }
+
+    /**
+     * A method for getting the list of packages for all levels of hierarchy.
+     *
+     * @return Array of packages of all hierarchies.
+     */
+    public final String[] getProjectList() {
+
+        /*
+         * If project list is already ready no need to construct again.
+         */
+        if (null != projectList)
+            return projectList;
+        hudson.tasks.junit.TestResult r = loadXml();
+        Collection<SuiteResult> suiteList = r.getSuites();
+
+        /*
+         * A set for the package names.
+         */
+        Map<Integer, HashSet<String>> projectSet = new HashMap<>();
+        int projectCount = 0;
+        final int projectCountLimit = 50;
+        int levelCap = Integer.MAX_VALUE;
+        for (SuiteResult suiteResult : suiteList) {
+            String suiteName = suiteResult.getName();
+            String projectName = "";
+
+            String[] packageTree = suiteName.split("[.]");
+
+            /*
+             * Iterating till the length-1 of packageTree array to exclude class name.
+             */
+            for (int level = 0; level < packageTree.length - 1 && level <= levelCap; level++) {
+                if (!projectName.isEmpty())
+                    projectName += '.';
+                projectName += packageTree[level];
+                if (!projectSet.containsKey(level)) {
+                    projectSet.put(level, new HashSet<String>());
+                }
+                if (!projectSet.get(level).contains(projectName)) {
+                    projectSet.get(level).add(projectName);
+                    projectCount++;
+                }
+                if (projectCount > projectCountLimit) {
+                    int levelToDel = projectSet.size() - 1;
+                    int deletionSize = projectSet.get(levelToDel).size();
+                    projectSet.remove(levelToDel);
+                    projectCount -= deletionSize;
+                    levelCap = projectSet.size() - 1;
+                }
+            }
+        }
+
+        /*
+         * Converting the set of package names to the corresponding array
+         * and finally sorting it in ascending order.
+         */
+        projectList = new String[projectCount];
```

Comment:
We can get rid of this array if we don't store the Hash instead just store the list with suite names.